src/timedata.cpp

```diff
@@ -11,107 +12,60 @@
-    // number of elements in vTimeOffsets is odd, which will never happen after
-    // there are 200 elements.
-    //
-    // But in this case the 'bug' is protective against some attacks, and may
-    // actually explain why we've never seen attacks which manipulate the
-    // clock offset.
-    //
-    // So we should hold off on fixing this and clean it up as part of
-    // a timing cleanup that strengthens it in a number of other ways.
-    //
-    if (vTimeOffsets.size() >= 5 && vTimeOffsets.size() % 2 == 1)
-    {
-        int64_t nMedian = vTimeOffsets.median();
-        std::vector<int64_t> vSorted = vTimeOffsets.sorted();
-        // Only let other nodes change our time by so much
-        if (abs64(nMedian) <= GetArg("-maxtimeadjustment", DEFAULT_MAX_TIME_ADJUSTMENT))
-        {
-            nTimeOffset = nMedian;
+    if (nPeersBehind + nPeersAhead < TIMEDATA_WARNING_SAMPLES) {
+        if (nTimeOffset < -TIMEDATA_WARNING_THRESHOLD) {
+            nPeersBehind++;
+        } else if (nTimeOffset > TIMEDATA_WARNING_THRESHOLD) {
+            nPeersAhead++;
         }
-        else
-        {
-            nTimeOffset = 0;
-
-            static bool fDone;
-            if (!fDone)
-            {
-                // If nobody has a time different than ours but within 5 minutes of ours, give a warning
-                bool fMatch = false;
-                BOOST_FOREACH(int64_t nOffset, vSorted)
-                    if (nOffset != 0 && abs64(nOffset) < 5 * 60)
-                        fMatch = true;
 
-                if (!fMatch)
-                {
-                    fDone = true;
-                    std::string strMessage = _("Warning: Please check that your computer's date and time are correct! If your clock is wrong Zcash will not work properly.");
-                    SetMiscWarning(strMessage);
-                    LogPrintf("*** %s\n", strMessage);
-                    uiInterface.ThreadSafeMessageBox(strMessage, "", CClientUIInterface::MSG_WARNING);
-                }
-            }
-        }
-        if (fDebug) {
-            BOOST_FOREACH(int64_t n, vSorted)
-                LogPrintf("%+d  ", n);
-            LogPrintf("|  ");
+        if (nPeersBehind + nPeersAhead == TIMEDATA_WARNING_SAMPLES) {
+            Warn(nPeersAhead, nPeersBehind);
         }
-        LogPrintf("nTimeOffset = %+d  (%+d minutes)\n", nTimeOffset, nTimeOffset/60);
     }
+    return nTimeOffset;
 }
+
+void CTimeWarning::Warn(size_t peersAhead, size_t peersBehind)
+{
+    std::string strMessage;
+    if (peersBehind >= TIMEDATA_WARNING_MAJORITY) {
+        strMessage = _("Warning: Your computer's date and time may be ahead of the rest of the network! If your clock is wrong Zcash will not work properly.");
+    } else if (peersAhead >= TIMEDATA_WARNING_MAJORITY) {
+        strMessage = _("Warning: Your computer's date and time may be behind the rest of the network! If your clock is wrong Zcash will not work properly.");
```

Comment:
Having separate strings is the right way to do it for internationalisation. (We don't currently translate any of the consensus failure messages, but the code is written so that we could.)

Reply:
oh, i see. thanks for the comment.